specs/chugsplash-config.md

```diff
@@ -0,0 +1,359 @@
+
+Valid contract references must reference contracts in the same ChugSplash project and use the following format:
+```
+{{ <Contract Reference Name> }}
+```
+
+### Example
+```js
+// Solidity
+Contract public contractReference;
+
+// Javascript
+const config: UserChugSplashConfig = {
+  // Configuration options for the project:
+  options: {
+    projectName: 'My Project',
+  },
+  contracts: {
+    Contract: {
+      contract: 'Contract',
+      ...
+    },
+    OtherContract: {
+      contract: 'SomeOtherContract',
+      variables: {
+        contractReference: '{{ Contract }}',
+      },
+    },
+  },
+}
+```
+
+## Strings
+For strings, we allow only unicode string literals.
+Note that if you supply a hex value in string format, we will not interpret it as a bytes value. We will encode it as if it were any other string.
+
+### Example
+```js
+// Solidity
+string public x;
+
+// Javascript
+x: 'literalString',
+```
+
+## User Defined Types
+We treat user defined types as if they were their underlying value type. We expect the input format to match the expected format of the underlying value type.
+
+### Example
+```js
+// Solidity
+type UserDefinedType is uint256;
+UserDefinedType public x;
+
+// Javascript
+x: 1,
+```
+
+## Enums
+Similar to user defined types, we treat enums like their underlying value type uint8. We accept the following possible input formats:
+- Base 10 integers between 0-255 where the integer corresponds to a valid Solidity enum value.
+- Base 10 integer strings between 0-255 where the integer corresponds to a valid Solidity enum value.
+- Typescript enums where the underlying number falls between 0-255 and corresponds to a valid Solidity enum value.
+
+### Examples
+```js
+// Solidity
+enum TestEnum { A, B, C }
+TestEnum public standardEnum;
+TestEnum public stringEnum;
+TestEnum public typescriptEnum;
+
+// Javascript Enum
+const enum TestEnum {
+  'A',
+  'B',
+  'C',
+}
+
+// Config vars
+standardEnum: 0,
+stringEnum: "1",
+typescriptEnum: TestEnum.C,
+```
+
+## Fixed-size Bytes Arrays
+For fixed sized bytes arrays, we only allow exactly 32 byte [DataHexStrings](https://docs.ethers.org/v5/api/utils/bytes/#DataHexString).
+
+### Examples
+```js
+// Solidity
+bytes32 public x;
+
+// Javascript
+x: '0x1111111111111111111111111111111111111111111111111111111111111111',
+```
+
+## Dynamically-sized Bytes
+For dynamic bytes, we accept arbitary length [DataHexStrings](https://docs.ethers.org/v5/api/utils/bytes/#DataHexString).
+
+### Example
+```js
+// Solidity
+bytes public x;
+
+// Javascript
+x: '0xabcd1234',
+```
+
+## Fixed-size Arrays
+For fixed length arrays, we accept JSON or JS arrays of exactly the defined length. The underlying array element types are validated individually and are expected to match the same input format as the corresponding value types.
```

Comment:
We currently allow users to define an empty array `[]` for fixed-size arrays. This is particularly useful for `__gap` variables, which are fixed-size arrays of length ~50 that appear in basically every upgradeable contract.
We have two choices:
1. Allow users to define `[]` for any fixed-size array. This is a footgun IMO because users should explicitly define each element in a fixed-size array the vast majority of the time.
2. Prevent users from defining `[]` for any fixed-size array. This is bad UX because users would need to write `__gap: [0, 0, ... 0]` for each gap variable. Also, since we currently create a SetStorage action for each config variable, we'd be creating roughly 50 SetStorage actions for each __gap variable.
3. Prevent users from defining `[]` for any fixed-size array _except_ `__gap` variables. Users simply cannot define `__gap` variables in their config.

I think 3 is the obvious choice. If you agree, we can leave this spec as is.

Reply:
We'll use a special keyword to resolve this which is an alias for a full fixed sized array full of zeros.